runtime/lua/vim/lsp.lua

```diff
@@ -411,6 +409,11 @@ local function on_client_exit(code, signal, client_id)
     end)
   end
 
+  vim.schedule(function()
+    local namespace = vim.lsp.diagnostic.get_namespace(client_id)
+    vim.diagnostic.reset(namespace)
+  end)
```

Comment:
I don't think you need a separate `vim.schedule` for this, it can go in the callback on line 418

Reply:
I think the above vim.schedule and loop could also be inverted:

```lua
  vim.schedule(function()
    for bufnr in pairs(client.attached_buffers) do
      if client and client.attached_buffers[bufnr] then
        api.nvim_exec_autocmds('LspDetach', {
          buffer = bufnr,
          modeline = false,
          data = { client_id = client_id },
        })
      end

      client.attached_buffers[bufnr] = nil

      if #lsp.get_clients({ bufnr = bufnr, _uninitialized = true }) == 0 then
        reset_defaults(bufnr)
      end
    end
    local namespace = vim.lsp.diagnostic.get_namespace(client_id)
    vim.diagnostic.reset(namespace)
  end)
```

Reply:
Good point. I moved it to the other `vim.schedule`.

Reply:
@mfussenegger actually, I like that better but I have to questions if I do that I should probaly also combine it with the final vim.schedule like this

```lua
  vim.schedule(function()
    for bufnr in pairs(client.attached_buffers) do
      if client and client.attached_buffers[bufnr] then
        api.nvim_exec_autocmds('LspDetach', {
          buffer = bufnr,
          modeline = false,
          data = { client_id = client_id },
        })
      end

      client.attached_buffers[bufnr] = nil

      if #lsp.get_clients({ bufnr = bufnr, _uninitialized = true }) == 0 then
        reset_defaults(bufnr)
      end
    end
    local namespace = vim.lsp.diagnostic.get_namespace(client_id)
    vim.diagnostic.reset(namespace)

    all_clients[client_id] = nil

    -- Client can be absent if executable starts, but initialize fails
    -- init/attach won't have happened
    if client then
      changetracking.reset(client)
    end
    if code ~= 0 or (signal ~= 0 and signal ~= 15) then
      local msg = string.format(
        'Client %s quit with exit code %s and signal %s. Check log for errors: %s',
        name,
        code,
        signal,
        lsp.get_log_path()
      )
      vim.notify(msg, vim.log.levels.WARN)
    end
  end)
``` 
I will check to see if that change passes the tests. If there isn't any reason we shouldn't combine all of these `vim.schedule`s, I will go ahead and make that change.

Reply:
Afaik the split into at least two schedule's is important to ensure the `LspDetach` events finished processing before the remainder of the cleanup runs.

Reply:
Although reading this comment before the last vim.schedule make me wonder if combining with the last vim.schedule could break autocommands using LspDetach. Any thoughts on if that is a possibility?

```  
 -- Schedule the deletion of the client object so that it exists in the execution of LspDetach
  -- autocommands
```

Reply:
> I will check to see if that change passes the tests.

Changing the code to below does pass the existing tests but I still am a bit concerned about if LspDetach autocommands will have access to the client so I am going to test it with my `minimal_init.lua` to see if it is actually safe. If it isn't I will switch to @mfussenegger original suggestion.

```lua
--- @param code integer
--- @param signal integer
--- @param client_id integer
local function on_client_exit(code, signal, client_id)
  local client = all_clients[client_id]

  -- Schedule the deletion of the client object so that it exists in the execution of LspDetach
  -- autocommands
  vim.schedule(function()
    for bufnr in pairs(client.attached_buffers) do
      if client and client.attached_buffers[bufnr] then
        api.nvim_exec_autocmds('LspDetach', {
          buffer = bufnr,
          modeline = false,
          data = { client_id = client_id },
        })
      end

      client.attached_buffers[bufnr] = nil

      if #lsp.get_clients({ bufnr = bufnr, _uninitialized = true }) == 0 then
        reset_defaults(bufnr)
      end
    end

    local name = client.name or 'unknown'

    all_clients[client_id] = nil

    -- Client can be absent if executable starts, but initialize fails
    -- init/attach won't have happened
    if client then
      changetracking.reset(client)
    end
    if code ~= 0 or (signal ~= 0 and signal ~= 15) then
      local msg = string.format(
        'Client %s quit with exit code %s and signal %s. Check log for errors: %s',
        name,
        code,
        signal,
        lsp.get_log_path()
      )
      vim.notify(msg, vim.log.levels.WARN)
    end
  end)
end
```

Reply:
I added the below LspDetach autocmd to my `minimal_init.lua` and got the same response "Client lua_ls is stopped" for the version of neovim with my code changes and also my unmodified version neovim on my system, so I believe it is a safe change.
```lua
vim.api.nvim_create_autocmd("LspDetach", {
	group = vim.api.nvim_create_augroup("lsp_detach", { clear = true }),
	desc = "lsp on_detach",
	callback = function(event)
		vim.notify("Detached Lsp client " .. event.data.client_id)
		local client = vim.lsp.get_client_by_id(event.data.client_id)

		if client == nil then
			vim.notify("Client is nil")
		elseif client.is_stopped() then
			vim.notify("Client " .. client.name .. " is stopped")
		else
			vim.notify("Client " .. client.name .. " is not stopped")
		end
	end,
})
```

Reply:
I want to test one more thing but will need to test it after work so I will convert this back to a draft PR.

Reply:
I'd still keep at least two schedule's for the backport to 0.10 to avoid risking a regression.
We can follow up with a refactor that puts all in one for 0.11

Reply:
Sounds good. I updated it to just use two schedules and moved the diagnostic reset up to the top one as it makes more since to be with that section of the code.

Reply:
I think this `vim.schedule` hack (?) should be commented, briefly talking about why we needed to defer its execution.

Reply:
@wookayin I think we can just make a followup PR right now to combine it all into one `vim.schedule` instead. @crwebb85 are you interested in doing that?